src/include/libpmemobj.h

```diff
@@ -84,6 +99,553 @@ void pmemobj_set_funcs(
+
+/*
+ * Allocates a new object with duplicate of the string s.
+ */
+PMEMoid pmemobj_strdup(PMEMobjpool *pop, const char *s, size_t n, int type_num);
+
+/*
+ * Frees an existing object.
+ */
+void pmemobj_free(PMEMoid oid);
+
+/*
+ * Returns the number of usable bytes in the object. May be greater than
+ * the requested size of the object because of internal alignment.
+ *
+ * Can be used with objects allocated by any of the available methods.
+ */
+size_t pmemobj_alloc_usable_size(PMEMoid oid);
+
+/*
+ * If called for the first time on a newly created pool, the root object
+ * of given size is allocated.  Otherwise, it returns the existing root object.
+ * In such case, the size must be not less than the actual root object size
+ * stored in the pool.  If it's larger, the root object is automatically
+ * resized.
+ *
+ * This function is thread-safe.
+ */
+PMEMoid pmemobj_get_root(PMEMobjpool *pop, size_t size);
+
+/*
+ * Returns the size in bytes of the root object. Always equal to the requested
+ * size.
+ */
+size_t pmemobj_sizeof_root(PMEMobjpool *pop);
+
+/*
+ * The following set of macros and functions allow access to the entire
+ * collection of objects, or objects of given type.
+ *
+ * Use with conjunction with non-transactional allocations. Pmemobj pool acts
+ * as a generic container (list) of objects that are not assigned to any
+ * user-defined data structures.
+ */
+
+/*
+ * Returns the first object of the specified type number.
+ */
+PMEMoid pmemobj_get_first(PMEMobjpool *pop, int type_num);
+
+/*
+ * Returns the next object (in order of allocations) of the same type.
+ */
+PMEMoid pmemobj_get_next(PMEMoid oid);
+
+/*
+ * Iterates through every existing allocated object.
+ */
+#define	PMEMOBJ_FOREACH(oidvar, vartype_num, pop)			\
+for (vartype_num = 0; vartype_num < MAX_OID_TYPE_NUM; ++vartype_num)	\
+	for (oidvar = pmemobj_get_first(pop, vartype_num);		\
+		(oidvar).off != 0; oidvar = pmemobj_get_next(oidvar))
+
+/*
+ * Iterates through every object of the specified type number.
+ */
+#define	PMEMOBJ_FOREACH_TYPE(var, pop, type_num)			\
+for (OID_ASSIGN(var, pmemobj_get_first(pop, type_num));			\
+		OID_IS_NULL(var) == 0;					\
+		OID_ASSIGN(var, pmemobj_get_next((var).oid)))
+
+/*
+ * Non-transactional persistent atomic circular doubly-linked list
+ */
+#define	PLIST_ENTRY(type)						\
+struct {								\
+	OID_TYPE(type) pe_next;						\
+	OID_TYPE(type) pe_prev;						\
+	PMEMmutex lock;							\
+}
+
+#define	PLIST_HEAD(name, type)						\
+struct name {								\
+	OID_TYPE(type) pe_first;					\
+	PMEMmutex lock;							\
+}
+
+int pmemobj_list_insert(PMEMobjpool *pop, size_t pe_offset, PMEMoid dest,
+	PMEMoid *head, int before, PMEMoid oid);
+
+int pmemobj_list_insert_new(PMEMobjpool *pop, size_t pe_offset, PMEMoid dest,
+	PMEMoid *head, int before, size_t size, int type_num);
+
+int pmemobj_list_remove(PMEMobjpool *pop, size_t pe_offset, PMEMoid oid,
+	PMEMoid *head, int free);
+
+int pmemobj_list_move(PMEMobjpool *pop,
```

Comment:
So, we can move element from one list to another, but only to the head or tail of the latter? Then, what is the value of having *insert* that allows to insert an element at the specific position in the list? I think it would be enough to insert at head or tail only. And since the lists will be circular, then head == tail, and in practice it doesn't really matter if you insert to head or tail.
So, I would rather simplify the *insert* functions:
```C
int pmemobj_list_insert(PMEMobjpool *pop, size_t pe_offset,
	PMEMoid *head, PMEMoid oid);

int pmemobj_list_insert_new(PMEMobjpool *pop, size_t pe_offset,
	PMEMoid *head, size_t size, int type_num);
```